ksp/src/test/kotlin/com/tschuchort/compiletesting/DelegatingSymbolProcessorRule.kt

```diff
@@ -0,0 +1,58 @@
+package com.tschuchort.compiletesting
+
+import org.jetbrains.kotlin.ksp.processing.CodeGenerator
+import org.jetbrains.kotlin.ksp.processing.Resolver
+import org.jetbrains.kotlin.ksp.processing.SymbolProcessor
+import org.junit.rules.TestWatcher
+import org.junit.runner.Description
+
+/**
+ * A processor implementation that delegates to an instance given in test.
+ * Must be used via [DelegatingSymbolProcessorRule] until we move this support into internals.
+ */
+internal class DelegateProcessor : SymbolProcessor {
+    private val delegate
+        get() = checkNotNull(Companion.delegate) {
+            "Delegate is not provided"
+        }
+
+    override fun finish() {
+        delegate.finish()
+    }
+
+    override fun init(options: Map<String, String>, kotlinVersion: KotlinVersion, codeGenerator: CodeGenerator) {
+        delegate.init(
+            options = options,
+            kotlinVersion = kotlinVersion,
+            codeGenerator = codeGenerator
+        )
+    }
+
+    override fun process(resolver: Resolver) {
+        delegate.process(resolver)
+    }
+
+    companion object {
+        var delegate: SymbolProcessor? = null
+    }
+}
+
+/**
+ * JUnit test rule that can trigger a single KSP processor.
+ * Useful for testing now until we figure out how to accept instance variables without changing the main
+ * [KotlinCompilation] class.
+ */
+class DelegatingSymbolProcessorRule : TestWatcher() {
+    fun delegateTo(processor: SymbolProcessor): Class<out SymbolProcessor> {
+        check(DelegateProcessor.delegate == null) {
+            "Cannot use DelegatingSymbolProcessorRule more than once"
+        }
+        DelegateProcessor.delegate = processor
+        return DelegateProcessor::class.java
+    }
+
+    override fun finished(description: Description?) {
+        DelegateProcessor.delegate = null
+        super.finished(description)
+    }
+}
```

Comment:
This is quite the hack if I may say so, but an interesting solution. What if the tests run concurrently? Maybe the `delegate` property should be thread local, but even then I think it would be possible to construct a case where you run into issues (using coroutines perhaps) and worst of all: If `K2JVMCompiler` ever switches threads, all bets are off. (Though the `ThreadLocal` hack for Kapt processors suffers from the same problem).

Reply:
yea i actually stole this idea from that thread local :)
for now this is only in our tests, not public.
I plan to followup w/ another CL where instance processors are supported.
Idk how to do that now but I think I'll byte-code genereate a delegating processor for each instance.
I tried it a bit but stopped when it got complicated.
https://github.com/yigit/kotlin-compile-testing/tree/ksp-with-instance-support